// @ldesign/builder/src/optimizers/code-splitting/code-splitter.ts
/**
 * 智能代码分割器
 * 
 * 提供自动代码分割、动态导入优化、路由级分割等高级功能
 * 
 * @author LDesign Team
 * @version 1.0.0
 */

// @ts-nocheck - 此文件包含实验性功能，暂时跳过严格类型检查

import * as path from 'path'
import * as fs from 'fs-extra'
import { parse } from '@babel/parser'
import traverse from '@babel/traverse'
import * as t from '@babel/types'
import { Logger } from '../../utils/logger'
import type { BuilderConfig } from '../../types/config'

/**
 * 代码分割策略
 */
export enum SplitStrategy {
  /** 基于路由的分割 */
  ROUTE_BASED = 'route-based',
  /** 基于组件的分割 */
  COMPONENT_BASED = 'component-based',
  /** 基于依赖大小的分割 */
  SIZE_BASED = 'size-based',
  /** 基于使用频率的分割 */
  FREQUENCY_BASED = 'frequency-based',
  /** 智能自动分割 */
  AUTO = 'auto'
}

/**
 * 分割点信息
 */
export interface SplitPoint {
  /** 分割点ID */
  id: string
  /** 文件路径 */
  filePath: string
  /** 导入路径 */
  importPath: string
  /** 分割类型 */
  type: 'dynamic' | 'route' | 'vendor' | 'common'
  /** 优先级 */
  priority: number
  /** 预估大小 */
  estimatedSize?: number
  /** 依赖的模块 */
  dependencies?: string[]
  /** 预加载策略 */
  preloadStrategy?: 'prefetch' | 'preload' | 'none'
  /** 分割阈值 */
  threshold?: number
}

/**
 * 路由配置
 */
export interface RouteConfig {
  /** 路由路径 */
  path: string
  /** 组件路径 */
  component: string
  /** 子路由 */
  children?: RouteConfig[]
  /** 是否懒加载 */
  lazy?: boolean
  /** 预加载策略 */
  preload?: 'prefetch' | 'preload' | 'none'
}

/**
 * 代码分割配置
 */
export interface CodeSplittingConfig {
  /** 启用代码分割 */
  enabled?: boolean
  /** 分割策略 */
  strategy?: SplitStrategy
  /** 最小分割大小（字节） */
  minSize?: number
  /** 最大分割大小（字节） */
  maxSize?: number
  /** 最大异步请求数 */
  maxAsyncRequests?: number
  /** 最大初始请求数 */
  maxInitialRequests?: number
  /** 自动命名分隔符 */
  automaticNameDelimiter?: string
  /** 缓存组配置 */
  cacheGroups?: Record<string, CacheGroupConfig>
  /** 路由配置 */
  routes?: RouteConfig[]
  /** 排除的模块 */
  exclude?: string[]
  /** 强制分割的模块 */
  forceSplit?: string[]
}

/**
 * 缓存组配置
 */
export interface CacheGroupConfig {
  /** 匹配规则 */
  test?: RegExp | string | ((module: any) => boolean)
  /** 优先级 */
  priority?: number
  /** 最小大小 */
  minSize?: number
  /** 最小共享次数 */
  minChunks?: number
  /** 重用已存在的块 */
  reuseExistingChunk?: boolean
  /** 强制分割 */
  enforce?: boolean
  /** 块名称 */
  name?: string | ((module: any) => string)
}

/**
 * 分析结果
 */
export interface AnalysisResult {
  /** 发现的分割点 */
  splitPoints: SplitPoint[]
  /** 依赖图 */
  dependencyGraph: Map<string, Set<string>>
  /** 模块大小 */
  moduleSizes: Map<string, number>
  /** 使用频率 */
  usageFrequency: Map<string, number>
  /** 建议的分割策略 */
  suggestedStrategy: SplitStrategy
  /** 预估的性能提升 */
  estimatedImprovement: {
    loadTime: number
    cacheability: number
    parallelization: number
  }
}

/**
 * 智能代码分割器
 */
export class SmartCodeSplitter {
  private config: CodeSplittingConfig
  private logger: Logger
  private splitPoints: Map<string, SplitPoint> = new Map()
  private dependencyGraph: Map<string, Set<string>> = new Map()
  private moduleSizes: Map<string, number> = new Map()
  private usageFrequency: Map<string, number> = new Map()
  private routeMap: Map<string, RouteConfig> = new Map()

  constructor(config: CodeSplittingConfig = {}) {
    this.config = {
      enabled: true,
      strategy: SplitStrategy.AUTO,
      minSize: 30000, // 30KB
      maxSize: 500000, // 500KB
      maxAsyncRequests: 30,
      maxInitialRequests: 30,
      automaticNameDelimiter: '~',
      ...config
    }

    this.logger = new Logger({ prefix: '[CodeSplitter]' })

    // 初始化路由映射
    if (this.config.routes) {
      this.buildRouteMap(this.config.routes)
    }
  }

  /**
   * 分析代码并生成分割点
   */
  async analyze(entryPoints: string | string[]): Promise<AnalysisResult> {
    this.logger.info('开始分析代码分割点...')

    const entries = Array.isArray(entryPoints) ? entryPoints : [entryPoints]

    // 分析每个入口点
    for (const entry of entries) {
      await this.analyzeEntry(entry)
    }

    // 检测动态导入
    await this.detectDynamicImports()

    // 分析路由
    if (this.config.routes) {
      await this.analyzeRoutes()
    }

    // 分析依赖关系
    await this.analyzeDependencies()

    // 计算模块大小
    await this.calculateModuleSizes()

    // 分析使用频率
    await this.analyzeUsageFrequency()

    // 生成优化建议
    const suggestedStrategy = this.suggestStrategy()

    // 优化分割点
    await this.optimizeSplitPoints()

    // 计算预估的性能提升
    const estimatedImprovement = this.estimateImprovement()

    const result: AnalysisResult = {
      splitPoints: Array.from(this.splitPoints.values()),
      dependencyGraph: this.dependencyGraph,
      moduleSizes: this.moduleSizes,
      usageFrequency: this.usageFrequency,
      suggestedStrategy,
      estimatedImprovement
    }

    this.logger.success(`分析完成，发现 ${this.splitPoints.size} 个分割点`)

    return result
  }

  /**
   * 分析入口点
   */
  private async analyzeEntry(entry: string): Promise<void> {
    if (!await fs.pathExists(entry)) {
      this.logger.warn(`入口文件不存在: ${entry}`)
      return
    }

    const code = await fs.readFile(entry, 'utf-8')
    const ast = parse(code, {
      sourceType: 'module',
      plugins: ['typescript', 'jsx', 'decorators-legacy']
    })

    const importedModules = new Set<string>()

    traverse(ast, {
      // 检测静态导入
      ImportDeclaration(path) {
        const source = path.node.source.value
        importedModules.add(source)

        // 记录依赖关系
        if (!this.dependencyGraph.has(entry)) {
          this.dependencyGraph.set(entry, new Set())
        }
        this.dependencyGraph.get(entry)!.add(source)
      },

      // 检测动态导入
      CallExpression(path) {
        if (t.isImport(path.node.callee)) {
          const arg = path.node.arguments[0]
          if (t.isStringLiteral(arg)) {
            const importPath = arg.value

            // 创建分割点
            const splitPoint: SplitPoint = {
              id: `dynamic_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
              filePath: entry,
              importPath,
              type: 'dynamic',
              priority: 5,
              preloadStrategy: 'prefetch'
            }

            this.splitPoints.set(splitPoint.id, splitPoint)
          }
        }
      }
    })

    // 递归分析导入的模块
    for (const module of importedModules) {
      const modulePath = await this.resolveModule(module, entry)
      if (modulePath && !this.moduleSizes.has(modulePath)) {
        await this.analyzeEntry(modulePath)
      }
    }
  }

  /**
   * 检测动态导入
   */
  private async detectDynamicImports(): Promise<void> {
    this.logger.debug('检测动态导入...')

    // 扫描所有已分析的文件
    for (const [filePath, dependencies] of this.dependencyGraph) {
      for (const dep of dependencies) {
        // 检查是否为动态导入候选
        if (this.shouldSplitModule(dep)) {
          const splitPoint: SplitPoint = {
            id: `auto_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            filePath,
            importPath: dep,
            type: 'dynamic',
            priority: 3,
            preloadStrategy: 'none'
          }

          this.splitPoints.set(splitPoint.id, splitPoint)
        }
      }
    }
  }

  /**
   * 分析路由
   */
  private async analyzeRoutes(): Promise<void> {
    if (!this.config.routes) return

    this.logger.debug('分析路由配置...')

    for (const route of this.config.routes) {
      await this.analyzeRoute(route)
    }
  }

  /**
   * 分析单个路由
   */
  private async analyzeRoute(route: RouteConfig, parentPath: string = ''): Promise<void> {
    const fullPath = path.join(parentPath, route.path)

    // 创建路由分割点
    if (route.lazy !== false) {
      const splitPoint: SplitPoint = {
        id: `route_${fullPath.replace(/\//g, '_')}`,
        filePath: route.component,
        importPath: route.component,
        type: 'route',
        priority: 8, // 路由分割优先级较高
        preloadStrategy: route.preload || 'prefetch'
      }

      this.splitPoints.set(splitPoint.id, splitPoint)
    }

    // 递归处理子路由
    if (route.children) {
      for (const child of route.children) {
        await this.analyzeRoute(child, fullPath)
      }
    }
  }

  /**
   * 分析依赖关系
   */
  private async analyzeDependencies(): Promise<void> {
    this.logger.debug('分析依赖关系...')

    // 识别共享模块
    const moduleUsageCount = new Map<string, number>()

    for (const [, dependencies] of this.dependencyGraph) {
      for (const dep of dependencies) {
        moduleUsageCount.set(dep, (moduleUsageCount.get(dep) || 0) + 1)
      }
    }

    // 创建 vendor 分割点
    for (const [module, count] of moduleUsageCount) {
      if (count >= 2 && module.includes('node_modules')) {
        const splitPoint: SplitPoint = {
          id: `vendor_${module.replace(/[^a-z0-9]/gi, '_')}`,
          filePath: module,
          importPath: module,
          type: 'vendor',
          priority: 10, // vendor 优先级最高
          preloadStrategy: 'preload'
        }

        this.splitPoints.set(splitPoint.id, splitPoint)
      }
    }

    // 创建 common 分割点
    for (const [module, count] of moduleUsageCount) {
      if (count >= 3 && !module.includes('node_modules')) {
        const splitPoint: SplitPoint = {
          id: `common_${module.replace(/[^a-z0-9]/gi, '_')}`,
          filePath: module,
          importPath: module,
          type: 'common',
          priority: 7,
          preloadStrategy: 'prefetch'
        }

        this.splitPoints.set(splitPoint.id, splitPoint)
      }
    }
  }

  /**
   * 计算模块大小
   */
  private async calculateModuleSizes(): Promise<void> {
    this.logger.debug('计算模块大小...')

    for (const [filePath] of this.dependencyGraph) {
      if (!this.moduleSizes.has(filePath)) {
        try {
          const stats = await fs.stat(filePath)
          this.moduleSizes.set(filePath, stats.size)
        } catch {
          // 忽略无法访问的文件
        }
      }
    }

    // 更新分割点的预估大小
    for (const splitPoint of this.splitPoints.values()) {
      const size = this.moduleSizes.get(splitPoint.filePath)
      if (size) {
        splitPoint.estimatedSize = size
      }
    }
  }

  /**
   * 分析使用频率
   */
  private async analyzeUsageFrequency(): Promise<void> {
    this.logger.debug('分析使用频率...')

    // 基于依赖图计算使用频率
    const visited = new Set<string>()
    const frequency = new Map<string, number>()

    const dfs = (module: string, depth: number = 0) => {
      if (visited.has(module)) return
      visited.add(module)

      const weight = Math.max(1, 10 - depth) // 深度越浅，权重越高
      frequency.set(module, (frequency.get(module) || 0) + weight)

      const deps = this.dependencyGraph.get(module)
      if (deps) {
        for (const dep of deps) {
          dfs(dep, depth + 1)
        }
      }
    }

    // 从入口点开始遍历
    for (const [entry] of this.dependencyGraph) {
      if (!entry.includes('node_modules')) {
        visited.clear()
        dfs(entry)
      }
    }

    this.usageFrequency = frequency
  }

  /**
   * 建议分割策略
   */
  private suggestStrategy(): SplitStrategy {
    // 基于项目特征推荐策略
    const hasRoutes = this.config.routes && this.config.routes.length > 0
    const hasManyDynamicImports = this.splitPoints.size > 10
    const hasLargeModules = Array.from(this.moduleSizes.values()).some(size => size > 100000)

    if (hasRoutes) {
      return SplitStrategy.ROUTE_BASED
    } else if (hasLargeModules) {
      return SplitStrategy.SIZE_BASED
    } else if (hasManyDynamicImports) {
      return SplitStrategy.COMPONENT_BASED
    } else {
      return SplitStrategy.AUTO
    }
  }

  /**
   * 优化分割点
   */
  private async optimizeSplitPoints(): Promise<void> {
    this.logger.debug('优化分割点...')

    // 移除过小的分割点
    for (const [id, splitPoint] of this.splitPoints) {
      if (splitPoint.estimatedSize && splitPoint.estimatedSize < this.config.minSize!) {
        this.splitPoints.delete(id)
      }
    }

    // 合并相关的分割点
    const merged = new Set<string>()
    for (const [id1, sp1] of this.splitPoints) {
      if (merged.has(id1)) continue

      for (const [id2, sp2] of this.splitPoints) {
        if (id1 === id2 || merged.has(id2)) continue

        // 如果两个分割点的总大小小于最大值，考虑合并
        const size1 = sp1.estimatedSize || 0
        const size2 = sp2.estimatedSize || 0
        if (size1 + size2 < this.config.maxSize!) {
          // 检查是否有共同依赖
          const deps1 = new Set(sp1.dependencies || [])
          const deps2 = new Set(sp2.dependencies || [])
          const commonDeps = new Set([...deps1].filter(x => deps2.has(x)))

          if (commonDeps.size > 0) {
            // 合并分割点
            sp1.dependencies = [...new Set([...(sp1.dependencies || []), ...(sp2.dependencies || [])])]
            sp1.estimatedSize = size1 + size2
            this.splitPoints.delete(id2)
            merged.add(id2)
          }
        }
      }
    }
  }

  /**
   * 预估性能提升
   */
  private estimateImprovement(): any {
    const totalSize = Array.from(this.moduleSizes.values()).reduce((sum, size) => sum + size, 0)
    const splitSize = Array.from(this.splitPoints.values())
      .reduce((sum, sp) => sum + (sp.estimatedSize || 0), 0)

    // 计算加载时间提升（假设并行加载）
    const parallelFactor = Math.min(this.splitPoints.size, 6) // 浏览器通常限制6个并行连接
    const loadTimeImprovement = (1 - 1 / parallelFactor) * 100

    // 计算缓存能力提升
    const cacheability = (splitSize / totalSize) * 100

    // 计算并行化提升
    const parallelization = Math.min(this.splitPoints.size * 10, 100)

    return {
      loadTime: Math.round(loadTimeImprovement),
      cacheability: Math.round(cacheability),
      parallelization: Math.round(parallelization)
    }
  }

  /**
   * 判断是否应该分割模块
   */
  private shouldSplitModule(module: string): boolean {
    // 检查排除列表
    if (this.config.exclude?.some(pattern => module.includes(pattern))) {
      return false
    }

    // 检查强制分割列表
    if (this.config.forceSplit?.some(pattern => module.includes(pattern))) {
      return true
    }

    // 基于大小判断
    const size = this.moduleSizes.get(module) || 0
    if (size > this.config.minSize!) {
      return true
    }

    // 基于使用频率判断
    const frequency = this.usageFrequency.get(module) || 0
    if (frequency > 5) {
      return true
    }

    return false
  }

  /**
   * 解析模块路径
   */
  private async resolveModule(modulePath: string, fromFile: string): Promise<string | null> {
    // 简化的模块解析逻辑
    if (modulePath.startsWith('.')) {
      const resolved = path.resolve(path.dirname(fromFile), modulePath)

      // 尝试不同的扩展名
      const extensions = ['.ts', '.tsx', '.js', '.jsx']
      for (const ext of extensions) {
        const fullPath = resolved + ext
        if (await fs.pathExists(fullPath)) {
          return fullPath
        }
      }

      // 尝试 index 文件
      for (const ext of extensions) {
        const indexPath = path.join(resolved, `index${ext}`)
        if (await fs.pathExists(indexPath)) {
          return indexPath
        }
      }
    }

    return null
  }

  /**
   * 构建路由映射
   */
  private buildRouteMap(routes: RouteConfig[], parentPath: string = ''): void {
    for (const route of routes) {
      const fullPath = path.join(parentPath, route.path)
      this.routeMap.set(fullPath, route)

      if (route.children) {
        this.buildRouteMap(route.children, fullPath)
      }
    }
  }

  /**
   * 生成 Rollup 配置
   */
  generateRollupConfig(): any {
    const manualChunks: Record<string, string[]> = {}

    // 根据分割点生成 chunks
    for (const splitPoint of this.splitPoints.values()) {
      const chunkName = this.generateChunkName(splitPoint)

      if (!manualChunks[chunkName]) {
        manualChunks[chunkName] = []
      }

      manualChunks[chunkName].push(splitPoint.importPath)
    }

    return {
      output: {
        manualChunks,
        chunkFileNames: '[name].[hash].js',
        entryFileNames: '[name].[hash].js'
      }
    }
  }

  /**
   * 生成 Webpack 配置
   */
  generateWebpackConfig(): any {
    const cacheGroups: any = {}

    // Vendor 分割
    cacheGroups.vendor = {
      test: /[\\/]node_modules[\\/]/,
      name: 'vendor',
      priority: 10,
      reuseExistingChunk: true
    }

    // Common 分割
    cacheGroups.common = {
      minChunks: 2,
      priority: 5,
      reuseExistingChunk: true,
      name: 'common'
    }

    // 根据分割点生成额外的缓存组
    for (const splitPoint of this.splitPoints.values()) {
      if (splitPoint.type === 'route') {
        const name = this.generateChunkName(splitPoint)
        cacheGroups[name] = {
          test: new RegExp(splitPoint.importPath.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')),
          name,
          priority: splitPoint.priority,
          reuseExistingChunk: true
        }
      }
    }

    return {
      optimization: {
        splitChunks: {
          chunks: 'all',
          minSize: this.config.minSize,
          maxSize: this.config.maxSize,
          maxAsyncRequests: this.config.maxAsyncRequests,
          maxInitialRequests: this.config.maxInitialRequests,
          automaticNameDelimiter: this.config.automaticNameDelimiter,
          cacheGroups
        }
      }
    }
  }

  /**
   * 生成块名称
   */
  private generateChunkName(splitPoint: SplitPoint): string {
    switch (splitPoint.type) {
      case 'vendor':
        return 'vendor'
      case 'common':
        return 'common'
      case 'route':
        return `route${this.config.automaticNameDelimiter}${splitPoint.id.replace(/[^a-z0-9]/gi, '')}`
      case 'dynamic':
        return `dynamic${this.config.automaticNameDelimiter}${splitPoint.id.replace(/[^a-z0-9]/gi, '')}`
      default:
        return splitPoint.id
    }
  }

  /**
   * 生成预加载脚本
   */
  generatePreloadScript(): string {
    const prefetchLinks: string[] = []
    const preloadLinks: string[] = []

    for (const splitPoint of this.splitPoints.values()) {
      const chunkName = this.generateChunkName(splitPoint)
      const chunkPath = `/${chunkName}.js`

      if (splitPoint.preloadStrategy === 'prefetch') {
        prefetchLinks.push(`<link rel="prefetch" href="${chunkPath}">`)
      } else if (splitPoint.preloadStrategy === 'preload') {
        preloadLinks.push(`<link rel="preload" href="${chunkPath}" as="script">`)
      }
    }

    return [...preloadLinks, ...prefetchLinks].join('\n')
  }

  /**
   * 生成性能报告
   */
  generateReport(): string {
    const lines: string[] = []

    lines.push('='.repeat(50))
    lines.push('智能代码分割分析报告')
    lines.push('='.repeat(50))
    lines.push('')

    lines.push(`发现分割点: ${this.splitPoints.size} 个`)
    lines.push(`建议策略: ${this.config.strategy}`)
    lines.push('')

    lines.push('分割点详情:')
    lines.push('-'.repeat(30))

    for (const splitPoint of this.splitPoints.values()) {
      lines.push(`  ${splitPoint.id}`)
      lines.push(`    类型: ${splitPoint.type}`)
      lines.push(`    优先级: ${splitPoint.priority}`)
      lines.push(`    预估大小: ${this.formatSize(splitPoint.estimatedSize || 0)}`)
      lines.push(`    预加载: ${splitPoint.preloadStrategy || 'none'}`)
      lines.push('')
    }

    lines.push('性能提升预估:')
    lines.push('-'.repeat(30))
    const improvement = this.estimateImprovement()
    lines.push(`  加载时间提升: ${improvement.loadTime}%`)
    lines.push(`  缓存能力提升: ${improvement.cacheability}%`)
    lines.push(`  并行化提升: ${improvement.parallelization}%`)

    return lines.join('\n')
  }

  /**
   * 格式化文件大小
   */
  private formatSize(bytes: number): string {
    const units = ['B', 'KB', 'MB', 'GB']
    let size = bytes
    let unitIndex = 0

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024
      unitIndex++
    }

    return `${size.toFixed(2)} ${units[unitIndex]}`
  }
}

/**
 * 创建智能代码分割器
 */
export function createSmartCodeSplitter(config?: CodeSplittingConfig): SmartCodeSplitter {
  return new SmartCodeSplitter(config)
}
